Support middleware for controller actions

alexpeachey opened the pull request:
@alexdavid 

Added support for putting a `middleware` key on your controller class.

If specified, it will add the listed middleware into the route handler chain.

```
middle = (req, res, next) ->
  req.custom = 'foo'
  next()

class UsersController
  middleware:
    index: middle
  
  index:   (req, res) -> res.end req.custom

module.exports = UsersController
```

Supports single or arrays of middleware for each entry. See feature for example of single and array usage.

alexdavid commented on src/controller_initializer.coffee:
```diff
@@ -32,5 +32,13 @@ class ControllerInitializer
+    return [] unless controller?
+    return [] unless typeof controller[actionName] is 'function'
+    return [] unless controller.middleware?[actionName]?
+    [].concat controller.middleware[actionName]
```
Awesome! So this means we can pass in an array of middleware instead of just a single one if needed?

alexpeachey replied:
Yeah, it takes a single or an array. I thought that would be a nicer API. You can see it in action in the feature specs.

alexdavid commented on test/feature/middleware_support.feature:
```diff
@@ -0,0 +1,58 @@
+    Examples:
+      | REQUEST | URL                  | RESPONSE BODY |
+      | GET     | /users               | foo           |
+
```
Can you add an extra empty new line here? I like having a double line break between scenarios

alexdavid commented on test/feature/middleware_support.feature:
```diff
@@ -0,0 +1,58 @@
+
+      class UsersController
+        middleware:
+          index: [middle1, middle2]
```
I love this, this is really cool!

alexdavid commented on src/controller_initializer.coffee:
```diff
@@ -32,5 +32,13 @@ class ControllerInitializer
       controllerName = camelCase Controller.name.replace /Controller$/, ''
       @controllers[controllerName] = new Controller @dependencies
 
```
I've been using 2 lines between methods

alexdavid commented:
This looks great @alexpeachey! It's a :ship: from me. @charlierudolph could you take a look at this as well and let me know what you think about it?

We should probably also document this new functionality. Either in this PR or I can open a new one and do it

alexpeachey commented:
I'll fix the spacing, and yes, I can document it. I wanted to make sure people were happy with it before updating the README.

charlierudolph commented on test/feature/middleware_support.feature:
```diff
@@ -0,0 +1,58 @@
+
+    Examples:
+      | REQUEST | URL                  | RESPONSE BODY |
+      | GET     | /users               | foo           |
```
The middle column can be shortened. Same for below

charlierudolph commented on test/feature/middleware_support.feature:
```diff
@@ -0,0 +1,58 @@
+        middleware:
+          index: middle
+
+        index:   (req, res) -> res.end req.custom
```
Please remove the spacing between the color and the first paren, same for below

charlierudolph commented:
Looks good to me

alexpeachey commented:
Fixed up the spacing issues and updated the README.
@alexdavid and @charlierudolph do you want to make sure the docs look ok before this gets shipped in?

charlierudolph commented on test/feature/middleware_support.feature:
```diff
@@ -0,0 +1,60 @@
+
+    Examples:
+      | REQUEST | URL     | RESPONSE BODY |
+      | GET     | /users  | foo           |
```
Ultra nitpick - 
There is still one extra space at the end of the middle column. Same as below.

Need to look into using [cucumber_lint](https://github.com/charlierudolph/cucumber_lint) for js projects. It automatically fixes tables formatting for you.

alexdavid replied:
Or you could use vim with [godlygeek/tabular](https://github.com/godlygeek/tabular) and these tables auto-align themselves to fit their content perfectly :)

(Props to @kevgo for showing me this awesome vim plugin)

charlierudolph commented:
Documentation looks good to me.

alexdavid commented:
:ship: from me

alexpeachey commented:
Because I originally cloned from my fork and then re-homed origin to here, I wasn't tracking this origin/master so my ship bombed out AFTER shipping, but before the cleanup. I'm closing this and deleting the branch.

charlierudolph commented:
If you add 

```
closes #<PR number>
```

into the commit summary it closes the PR for you and creates a link to the commit
